### PR TITLE
Provide Tensor Padding Helpers #960

### LINKED ISSUE #960: Provide Tensor Padding Helpers
### Feature description

Would be great to have helpers in Tensor for padding. Similar to PyTorch's [pad](https://pytorch.org/docs/stable/generated/torch.nn.functional.pad.html) and Candle's [pad_with_zeros](https://github.com/huggingface/candle/blob/c6763e3b41790bea0459d3838c3ed1d59d5174ef/candle-core/src/tensor.rs#L2246)

### Additional Information

Some API like this would likely be good (result of conversation on Discord):

```rust
pub fn pad<const D: usize>(self, pad: Padding, pad_mode: PadMode, value: K::Elem) -> Tensor<B, D, K>


// Possible definition of padding
pub struct Padding {
    left: usize,
    right: usize,
    top: usize,
    bottom: usize,
}
impl Padding {
    pub fn uniform(value: usize) -> Self {
        Self {
            left: value,
            right: value,
            top: value,
            bottom: value,
        }
    }
    
    // Additional helpers to create useful padding configs
}
```

### PR DESCRIPTION
**Pull request by jcmullwh**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x]  Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

Fixes #960

### Changes

This PR addresses the need for padding functionality in tensors of float and integer types. This is an initial implementation covering only constant value padding. The solution implemented involves adding methods for both constant padding mode and handling different padding sizes. The changes enhance the library's functionality by enabling users to easily manipulate tensor dimensions. The functionality is largely based on pytorch listed as an example.

### Testing

The testing for these changes involves unit tests added to the tensor API testing module. These tests cover multiple scenarios including padding 2D and 4D tensors with constant values, and applying asymmetric padding. The tests ensure the correct implementation of padding across different tensor types and dimensions. All new tests have been executed successfully with the existing test suite, confirming that the new feature works as expected.

### Remaining Work
#1535

### CODE REVIEW

File: burn-tensor/src/tensor/api/base.rs [hunk 1]
```diff
@@ -530,6 +530,184 @@ where
...
+                    .try_into()
+                    .unwrap();
+
+                // Create the padded tensor
+                let padded_shape = Shape::from(padded_dims);
+                let padded_data = Data::full(padded_shape, padding_value);
+                let padded_tensor = Tensor::<B, D, K>::from_data_devauto(padded_data);
+
+                // Assign the original tensor data to the appropriate slice of the padded tensor
+                padded_tensor.slice_assign(ranges, self)
```
Comment (nathanielsimard) [inline burn-tensor/src/tensor/api/base.rs:612]: Using `Data` is not efficient; it's preferable to use `Tensor::full(shape, value)` instead of `Data::full`.

File: burn-tensor/src/tensor/api/base.rs [hunk 2]
```diff
@@ -530,6 +530,184 @@ where
...
+    ///    let padded_tensor = tensor.pad(padding, PadMode::Constant, Some(pad_const.into()));
+    /// }    
+    ///
+    ///
+    /// ```
+    /// Pads the tensor with the specified padding configuration and padding mode.
+    pub fn pad(
+        self,
+        pad_amount: Padding,
+        pad_mode: PadMode,
```
Comment (nathanielsimard) [inline burn-tensor/src/tensor/api/base.rs:572]: I would change a bit the API

`Padding` => `PadSize`.

The mode should contains the value:

```rust
let tensor = tensor.pad(PadSize::uniform(2), PadMode::Constant(0));
```

File: burn-tensor/src/tensor/api/base.rs [hunk 3]
```diff
@@ -530,6 +530,184 @@ where
...
+pub enum PadMode {
+    /// Pads the tensor with a constant value.
+    Constant,
+    // Reflect,
+    // Replicate,
+    // Other modes can be added here
+}
+
+/// Represents padding quantity for a tensor.
+pub struct Padding {
```
Comment (nathanielsimard) [inline burn-tensor/src/tensor/api/base.rs:628]: `PadSize`

**Review by nathanielsimard — CHANGES_REQUESTED** [review]
Thanks for the contribution! I primarily reviewed the API, but there are still some fixes to be made regarding remaining prints and comments.

File: burn-tensor/src/tensor/api/base.rs [hunk 4]
```diff
@@ -530,6 +530,177 @@ where
...
+    ///
+    ///
+    /// ```
+
+    pub fn pad(self, pad_size: PadSize, pad_mode: PadMode<K, B>) -> Tensor<B, D, K>
+    where
+        K::Elem: Element,
+        K: Numeric<B>,
+    {
+        match pad_mode {
```
Comment (nathanielsimard) [inline burn-tensor/src/tensor/api/base.rs:572]: I think this function should be defined in the file `numeric`, you won't have to specify any `where`.

File: burn-tensor/src/tensor/api/base.rs [hunk 5]
```diff
@@ -530,6 +530,177 @@ where
...
+                            0..dim
+                        }
+                    })
+                    .collect::<Vec<Range<usize>>>()
+                    .try_into()
+                    .unwrap();
+
+                // Create the padded tensor
+                let padded_shape = Shape::from(padded_dims);
+                let padded_tensor = Tensor::full_devauto(padded_shape, value);
```
Comment (nathanielsimard) [inline burn-tensor/src/tensor/api/base.rs:599]: You can merge with `main` and add a device argument to avoid using the default device.

File: burn-tensor/src/tensor/api/base.rs [hunk 6]
```diff
@@ -530,6 +530,177 @@ where
...
+pub enum PadMode<K, B>
+where
+    B: Backend,
+    K: BasicOps<B>,
+    K::Elem: Element,
+{
+    /// Pads the tensor with a constant value.
+    Constant(K::Elem),
+    // Other modes can be added here
+}
```
Comment (nathanielsimard) [inline burn-tensor/src/tensor/api/base.rs:618]: For the definition, I would simplify it with just use `E: Element` as generic argument.

**Review by nathanielsimard — CHANGES_REQUESTED** [review]
Just some minor comments to be updated with the new changes added to the API.

**Comment by antimora** [comment]
Let us know if you need more assistance with this PR.

**Comment by jcmullwh** [comment]
> Let us know if you need more assistance with this PR.

Got bogged down the last couple weeks, should be able to get it fixed this weekend!

**Comment by antimora** [comment]
@nathanielsimard, I have cleaned up this PR and resolved all the issues so we can merge it. The remaining work (bool tensor) is in a separate ticket https://github.com/tracel-ai/burn/issues/1535 (assigned to you)

File: crates/burn-tensor/src/tensor/ops/bool_tensor.rs [hunk 7]
```diff
@@ -463,4 +463,16 @@ pub trait BoolTensorOps<B: Backend> {
...
+    fn bool_full<const D: usize>(
+        shape: Shape<D>,
+        value: bool,
+        device: &Device<B>,
+    ) -> BoolTensor<B, D> {
+        B::int_equal_elem(
+            B::int_zeros(shape, device),
+            if value { 0.elem() } else { 1.elem() },
+        )
+    }
```
Comment (nathanielsimard) [inline crates/burn-tensor/src/tensor/ops/bool_tensor.rs:477]: I guess we could actually remove this from the PR?

File: crates/burn-tensor/src/tensor/api/numeric.rs [hunk 8]
```diff
@@ -736,6 +738,49 @@ where
...
+    /// # Arguments
+    ///
+    /// * `pads` - A tuple of four integers representing the padding on the left, right, top, and bottom.
+    /// * `value` - The value to pad the tensor with.
+    ///
+    /// # Returns
+    ///
+    /// A new tensor with the given padding.
+    pub fn pad(self, pads: (usize, usize, usize, usize), value: K::Elem) -> Tensor<B, D, K> {
+        let (left, right, top, bottom) = pads;
```
Comment (laggui) [inline crates/burn-tensor/src/tensor/api/numeric.rs:753]: Small nitpick: can we rename `pads` to `padding` (or even `pad` could work, the plural just sounds weird)?

**Review by laggui — APPROVED** [review]
LGTM! Apart from the existing review comment I only have one minor comment regarding parameter naming but I'm approving in advance.

File: crates/burn-tensor/src/tensor/ops/bool_tensor.rs (hunk 7 above)
Comment (antimora) [reply crates/burn-tensor/src/tensor/ops/bool_tensor.rs:477]: OK. I have moved to #1535 preserve the code.

File: crates/burn-tensor/src/tensor/api/numeric.rs (hunk 8 above)
Comment (antimora) [reply crates/burn-tensor/src/tensor/api/numeric.rs:753]: Renamed to padding

**Review by nathanielsimard — APPROVED** [review]
